File: fc-chat/demo/part3/ch2/02-security-login-logout/src/main/java/net/prostars/messagesystem/config/SecurityConfig.java
package net.prostars.messagesystem.config;

import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import java.io.IOException;
import net.prostars.messagesystem.auth.RestApiLoginAuthFilter;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.ProviderManager;
import org.springframework.security.authentication.dao.DaoAuthenticationProvider;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.configurers.AbstractHttpConfigurer;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.userdetails.User;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.security.provisioning.InMemoryUserDetailsManager;
import org.springframework.security.web.SecurityFilterChain;
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter;
import org.springframework.security.web.util.matcher.AntPathRequestMatcher;

@Configuration
@SuppressWarnings("unused")
public class SecurityConfig {

  private static final Logger log = LoggerFactory.getLogger(SecurityConfig.class);

  @Bean
  public PasswordEncoder passwordEncoder() {
    return new BCryptPasswordEncoder();
  }

  @Bean
  public UserDetailsService userDetailsService() {
    UserDetails userDetails =
        User.builder()
            .username("testuser")
            .password(passwordEncoder().encode("testpass"))
            .roles("USER")
            .build();
    return new InMemoryUserDetailsManager(userDetails);
  }

  @Bean
  public AuthenticationManager authenticationManager(
      UserDetailsService detailsService, PasswordEncoder passwordEncoder) {
    DaoAuthenticationProvider daoAuthenticationProvider = new DaoAuthenticationProvider();
    daoAuthenticationProvider.setUserDetailsService(detailsService);
    daoAuthenticationProvider.setPasswordEncoder(passwordEncoder);
    return new ProviderManager(daoAuthenticationProvider);
  }

  @Bean
  public SecurityFilterChain securityFilterChain(
      HttpSecurity httpSecurity, AuthenticationManager authenticationManager) throws Exception {
    RestApiLoginAuthFilter restApiLoginAuthFilter =
        new RestApiLoginAuthFilter(
            new AntPathRequestMatcher("/api/v1/auth/login", "POST"), authenticationManager);

    httpSecurity
        .csrf(AbstractHttpConfigurer::disable)
        .httpBasic(AbstractHttpConfigurer::disable)
        .formLogin(AbstractHttpConfigurer::disable)
        .addFilterAt(restApiLoginAuthFilter, UsernamePasswordAuthenticationFilter.class)
        .authorizeHttpRequests(
            auth ->
                auth.requestMatchers("/api/v1/auth/login").permitAll().anyRequest().authenticated())
        .logout(
            logout ->
                logout.logoutUrl("/api/v1/auth/logout").logoutSuccessHandler(this::logoutHandler));
    return httpSecurity.build();
  }

  private void logoutHandler(
      HttpServletRequest request, HttpServletResponse response, Authentication authentication) {
    response.setContentType(MediaType.TEXT_PLAIN_VALUE);
    response.setCharacterEncoding("UTF-8");
    String message;

    if (authentication != null && authentication.isAuthenticated()) {
      response.setStatus(HttpStatus.OK.value());
      message = "Logout success.";
    } else {
      response.setStatus(HttpStatus.UNAUTHORIZED.value());
      message = "Logout failed.";
    }

    try {
      response.getWriter().write(message);
    } catch (IOException ex) {
      log.error("Response failed. cause: {}", ex.getMessage());
    }
  }
}

File: fc-chat/demo/part3/with-push/src/main/kotlin/com/yscorp/withpush/messagesystem/config/SecurityConfig.kt
package com.yscorp.withpush.messagesystem.config

import com.yscorp.withpush.messagesystem.auth.RestApiLoginAuthFilter
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.authentication.ProviderManager
import org.springframework.security.authentication.dao.DaoAuthenticationProvider
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.config.annotation.web.configurers.*
import org.springframework.security.core.Authentication
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.security.web.SecurityFilterChain
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter
import org.springframework.security.web.util.matcher.AntPathRequestMatcher
import java.io.IOException

@Configuration
@Suppress("unused")
class SecurityConfig {

    @Bean
    fun passwordEncoder(): PasswordEncoder {
        return BCryptPasswordEncoder()
    }

    @Bean
    fun authenticationManager(
        detailsService: UserDetailsService, passwordEncoder: PasswordEncoder?
    ): AuthenticationManager {
        val daoAuthenticationProvider = DaoAuthenticationProvider()
        daoAuthenticationProvider.setUserDetailsService(detailsService)
        daoAuthenticationProvider.setPasswordEncoder(passwordEncoder)

        return ProviderManager(daoAuthenticationProvider)
    }

    @Bean
    @Throws(Exception::class)
    fun securityFilterChain(
        httpSecurity: HttpSecurity, authenticationManager: AuthenticationManager
    ): SecurityFilterChain {
        val restApiLoginAuthFilter = RestApiLoginAuthFilter(
                AntPathRequestMatcher("/api/v1/auth/login", "POST"), authenticationManager
            )

        httpSecurity
            .csrf { obj: CsrfConfigurer<HttpSecurity?> -> obj.disable() }
            .httpBasic { obj: HttpBasicConfigurer<HttpSecurity?> -> obj.disable() }
            .formLogin { obj: FormLoginConfigurer<HttpSecurity?> -> obj.disable() }
            .addFilterAt(restApiLoginAuthFilter, UsernamePasswordAuthenticationFilter::class.java)
            .authorizeHttpRequests { auth ->
                auth.requestMatchers("/api/v1/auth/register", "/api/v1/auth/login")
                    .permitAll()
                    .anyRequest()
                    .authenticated()
            }
            .logout { logout: LogoutConfigurer<HttpSecurity?> ->
                logout.logoutUrl("/api/v1/auth/logout").logoutSuccessHandler { request: HttpServletRequest, response: HttpServletResponse, authentication: Authentication? ->
                    this.logoutHandler(
                        request,
                        response,
                        authentication
                    )
                }
            }
        return httpSecurity.build()
    }

    private fun logoutHandler(
        request: HttpServletRequest, response: HttpServletResponse, authentication: Authentication?
    ) {
        response.contentType = MediaType.TEXT_PLAIN_VALUE
        response.characterEncoding = "UTF-8"
        val message: String

        if (authentication != null && authentication.isAuthenticated) {
            response.status = HttpStatus.OK.value()
            message = "Logout success."
        } else {
            response.status = HttpStatus.UNAUTHORIZED.value()
            message = "Logout failed."
        }

        try {
            response.writer.write(message)
        } catch (ex: IOException) {
            log.error("Response failed. cause: {}", ex.message)
        }
    }

    companion object {
        private val log: Logger = LoggerFactory.getLogger(SecurityConfig::class.java)
    }
}
